clients/command/src/main/java/org/eclipse/hono/client/command/CommandRouterCommandConsumerFactory.java

```diff
@@ -288,4 +354,42 @@ private Future<Void> removeCommandConsumer(
                     }
                 });
     }
+
+    private Future<Void> handleClosingConsumers(final SpanContext onReleaseSpanContext) {
+        if (pendingConsumersToClose.size() >= batchSize) {
+            return removeCommandConsumers(pendingConsumersToClose.size(), onReleaseSpanContext);
+        } else {
+            // start the timer if not started
+            if (batchTimerId == -1) {
+                final Vertx vertx = Vertx.currentContext().owner();
+                if (vertx != null) {
+                    batchTimerId = vertx.setTimer(batchMaxTimeout, aLong -> {
+                        final Span span = TracingHelper
+                                .buildSpan(tracer, null, "on consumers timer", References.CHILD_OF)
+                                .ignoreActiveSpan()
+                                .withTag(Tags.COMPONENT.getKey(), "CommandConsumerTimer")
+                                .start();
+                        removeCommandConsumers(pendingConsumersToClose.size(), span.context())
+                                .onComplete(v -> span.finish());
+                        // mark the timer as finished although the batch operation is still ongoing
+                        batchTimerId = -1;
+                    });
+                }
+            }
+            return Future.succeededFuture();
+        }
+    }
+
+    private Future<Void> removeCommandConsumers(final int count, final SpanContext spanContext) {
```

Comment:
Currently, `count` is always set to `pendingConsumersToClose.size()` here, making the parameter look superfluous.

Just looking at the code of this method, it also should better be made obvious that `pendingConsumersToClose.remove(0)` below can't produce an `IndexOutOfBoundsException`.

Reply:
I will improve.
At least the first call should use batchSize instead of pendingConsumersToClose.size() as first parameter.